### PR TITLE
Clean up InputBuffer

### PR DESCRIPTION
**Pull request by apaszke**
Also accept sparse tensors or corresponding types in VariableType.

Note that this is expected to fail until we merge the new interpreter, because it uncovers bugs that were fixed in that PR.

### CODE REVIEW

File: torch/csrc/autograd/input_buffer.cpp [hunk 1]
```diff
@@ -6,43 +6,36 @@
...
-  if (!item.first.defined()) {
-    auto current_version = var.current_version();
-    buffer[pos] = std::make_pair<>(std::move(var), current_version);
+  auto& old_var = buffer[pos];
+  if (!old_var.defined()) {
+    buffer[pos] = std::move(var);
   } else {
-    auto result = apply_fn<Add>()(item.first, std::move(var));
-    buffer[pos] = std::make_pair<>(std::move(result), 0);
+    // ATen doesn't route sparse additions correctly...
```
Comment (ezyang) [inline torch/csrc/autograd/input_buffer.cpp]: File a bug.

**Comment by ezyang** [comment]
Looks fine, but I couldn't tell from the commit message why we're allowed to drop the version information.

**Review by zdevito — APPROVED** [review]
ATen changes look fine. Not for this PR: eventually we probably want to add toSparse(...) to Type so that things like VariableType can be handled correctly.

**Review by colesbury — APPROVED** [review]
lgtm

**Comment by apaszke** [comment]
Don't merge yet. The comment saying that "ATen doesn't dispatch the method correctly" was really VariableType.cpp that didn't allow sparse types in `checked_cast`. I'll try to simplify that part of the code and see if it fails (and yes, will file a bug report otherwise @ezyang).

@ezyang The version is not needed because it's not used anywhere (look at the full code before the changes)

**Pull request merged**